### PR TITLE
DRAFT: Fix failing physx tests due to bad timeouts

### PR DESCRIPTION
**Pull request by FuzzyCarterAWS**
**General Automation Improvements**
* Adding Physics Constant Utils file and supporting __init__ files to enable referencing in tests

**#9365**
* Refactored tests for usage Physics Constants
* Increased timeout to 3 seconds after determining that a HIGH SPEC machine gets close to a 1 second timeout

**#9368** 
* STILL FIXING

**Validation**
* Verified tests passed locally 5x in a row

### CODE REVIEW

File: AutomatedTesting/Gem/PythonTests/Physics/utils/physics_constants.py
Comment (hultonha) [inline AutomatedTesting/Gem/PythonTests/Physics/utils/physics_constants.py:12]: update: GetWorldTranslation is the `TransformBus`, might be worth moving to its own section

File: AutomatedTesting/Gem/PythonTests/Physics/utils/physics_constants.py
Comment (hultonha) [inline AutomatedTesting/Gem/PythonTests/Physics/utils/physics_constants.py:24]: uber nit: Missing EOL character

**Review by hultonha — COMMENTED** [review]
I think this is okay but are we just really increasing the timeout? Is there another way we can resolve this without relying on waiting longer, as it seems this could still mean the test might fail in certain circumstances. Is there an event or notification we use instead (apologies if I'm missing something as I'm not as familiar with python tests overall)

File: AutomatedTesting/Gem/PythonTests/Physics/tests/force_region/ForceRegion_NoQuiverOnHighLinearDampingForce.py
Comment (smurly) [inline AutomatedTesting/Gem/PythonTests/Physics/tests/force_region/ForceRegion_NoQuiverOnHighLinearDampingForce.py:93]: seems to me like the entire class Entity here should be in a helper as class PhysicsEntity(EditorEntity):
EditorEntity from editor_entity_utils.py
this gives you lots of basic entity utilities and then you can add all these reusable physics functions and more.
usage would be slightly different in that you would then call PhysicsEntity.find_editor_entity(name) to construct a rather than the current Entity(name).
just a though not saying this needs changing

File: AutomatedTesting/Gem/PythonTests/Physics/utils/physics_constants.py
Comment (smurly) [inline AutomatedTesting/Gem/PythonTests/Physics/utils/physics_constants.py:24]: nit: newline at end of file

**Review by smurly — APPROVED** [review]
physics.RigidBodyRequestBus is just as likely to change as "GetLinearVelocity". putting the helper functions in a utility module would centralize them where the literal strings would be in one place just like a constants file, but also address the issue of bus names. They're already in a class which is good, but why only in this test module?

File: AutomatedTesting/Gem/PythonTests/Physics/tests/force_region/ForceRegion_NoQuiverOnHighLinearDampingForce.py
Comment (FuzzyCarterAWS) [reply AutomatedTesting/Gem/PythonTests/Physics/tests/force_region/ForceRegion_NoQuiverOnHighLinearDampingForce.py:93]: This is something I considered too. 

I'm looking at these tests and see great refactoring opportunities to improve the code throughout the test code base, but I also don't want to go down the rabbit hole quite yet since there's other refactors of the physics code already being considered where this could be added in.

**Comment by FuzzyCarterAWS** [comment]
@hultonha 
> I think this is okay but are we just really increasing the timeout? Is there another way we can resolve this without relying on waiting longer, as it seems this could still mean the test might fail in certain circumstances. Is there an event or notification we use instead (apologies if I'm missing something as I'm not as familiar with python tests overall)

The wait here we're increasing is a "wait until" lambda event timeout that waits for one of two conditions, the lambda event to finish (pass) or the timeout to be hit (fail). We will only wait as long as it takes the event to be returned if things are operating as expected. Whatever methodology we use to detect if the action has completed, we'll want to put in some sore of timeout to investigate why things are taking too long. In this case, through local testing on a HIGH SPEC machine, I noticed that I could get over .8 seconds occasionally, which means that a LOW/MID SPEC machine could spike even higher. 

At some point, we need to add a cutoff and re-approach our test design, and in this case I believe anything over 3 seconds should be that case. Your quest here however has made it evident that I need to update the comments in front of these waits to denote that so we don't continuously bump times up to accommodate slow executing tests. In the end, there's always going to be some time spent waiting for the simulation to fire off events to be detected, simply due to the nature that the simulation actually takes time to execute.

File: AutomatedTesting/Gem/PythonTests/Physics/tests/force_region/ForceRegion_NoQuiverOnHighLinearDampingForce.py
Comment (smurly) [reply AutomatedTesting/Gem/PythonTests/Physics/tests/force_region/ForceRegion_NoQuiverOnHighLinearDampingForce.py:93]: editor_entity_utils.py EditorEntity already has transform bus helpers. you might even be able to just add physics bus to it.
    def get_world_translation(self) -> azlmbr.math.Vector3:
        """
        Gets the world translation of the entity
        """
        return azlmbr.components.TransformBus(azlmbr.bus.Event, "GetWorldTranslation", self.id)

**Comment by greerdv** [comment]
> @hultonha
> 
> > I think this is okay but are we just really increasing the timeout? Is there another way we can resolve this without relying on waiting longer, as it seems this could still mean the test might fail in certain circumstances. Is there an event or notification we use instead (apologies if I'm missing something as I'm not as familiar with python tests overall)
> 
> The wait here we're increasing is a "wait until" lambda event timeout that waits for one of two conditions, the lambda event to finish (pass) or the timeout to be hit (fail). We will only wait as long as it takes the event to be returned if things are operating as expected. Whatever methodology we use to detect if the action has completed, we'll want to put in some sore of timeout to investigate why things are taking too long. In this case, through local testing on a HIGH SPEC machine, I noticed that I could get over .8 seconds occasionally, which means that a LOW/MID SPEC machine could spike even higher.
> 
> At some point, we need to add a cutoff and re-approach our test design, and in this case I believe anything over 3 seconds should be that case. Your question here however has made it evident that I need to update the comments in front of these waits to denote that so we don't continuously bump times up to accommodate slow executing tests. In the end, there's always going to be some time spent waiting for the simulation to fire off events to be detected, simply due to the nature that the simulation actually takes time to execute.

Yeah, we should be changing tests to complete faster rather than extending timeouts. I looked at one test today which drops a ball 8 metres and waits for it to hit the ground (which should take around 1.3s). We can easily make the drop a lot shorter (even 0.1m) and still test the same functionality.

**Comment by hultonha** [comment]
> > @hultonha
> > > I think this is okay but are we just really increasing the timeout? Is there another way we can resolve this without relying on waiting longer, as it seems this could still mean the test might fail in certain circumstances. Is there an event or notification we use instead (apologies if I'm missing something as I'm not as familiar with python tests overall)
> > 
> > 
> > The wait here we're increasing is a "wait until" lambda event timeout that waits for one of two conditions, the lambda event to finish (pass) or the timeout to be hit (fail). We will only wait as long as it takes the event to be returned if things are operating as expected. Whatever methodology we use to detect if the action has completed, we'll want to put in some sore of timeout to investigate why things are taking too long. In this case, through local testing on a HIGH SPEC machine, I noticed that I could get over .8 seconds occasionally, which means that a LOW/MID SPEC machine could spike even higher.
> > At some point, we need to add a cutoff and re-approach our test design, and in this case I believe anything over 3 seconds should be that case. Your question here however has made it evident that I need to update the comments in front of these waits to denote that so we don't continuously bump times up to accommodate slow executing tests. In the end, there's always going to be some time spent waiting for the simulation to fire off events to be detected, simply due to the nature that the simulation actually takes time to execute.
> 
> Yeah, we should be changing tests to complete faster rather than extending timeouts. I looked at one test today which drops a ball 8 metres and waits for it to hit the ground (which should take around 1.3s). We can easily make the drop a lot shorter (even 0.1m) and still test the same functionality.

Thanks for the additional information @FuzzyCarterAWS and @greerdv - I think it's likely a balance of the two with a compromise somewhere in the middle. I'm definitely in favour of updating the tests to simply complete more quickly but understand a small tolerance increase on the timeouts may be prudent

**Comment by FuzzyCarterAWS** [comment]
@hultonha We've been discussing replacing the physics test levels with programmatically generated test levels from the scripts themselves.  I agree that we should be improving simulation execution times, but with the knowledge that we want to do the above, I believe we can approach those simulation execution time test design approaches at that time of updating the tests to follow that paradigm. If we put the effort into doing that prematurely, we will spend more engineering time redesigning the level for the existing test paradigm and then have to do it once again when we do them programmatically.

However, with @greerdv's work in charting times for the wait_until calls he's evaluated, we can expand upon that and prioritize tests that we see are running long & approaching the timeout first.

In the mean time, bumping the timeout to get stable and executing tests, is the priority call I would make.

**Comment by hultonha** [comment]
Let me know when you need another review of this @FuzzyCarterAWS - also as @moraaar mentioned in sig-simulation triage, it would be good to reproduce these tests repeatedly on Linux before merging as the failures seem to be more prevalent there than on Windows

**Comment by Kadino** [comment]
Closing stale PR, please reopen when seeking new feedback

**Pull request closed without merging**
